src/lib/optimizer/strategy/disjunction_to_union_rule.cpp

```diff
@@ -0,0 +1,63 @@
+#include "disjunction_to_union_rule.hpp"
+
+#include <expression/expression_utils.hpp>
+
+#include "expression/logical_expression.hpp"
+#include "logical_query_plan/lqp_utils.hpp"
+#include "logical_query_plan/predicate_node.hpp"
+#include "logical_query_plan/union_node.hpp"
+
+namespace opossum {
+
+std::string DisjunctionToUnionRule::name() const { return "Disjunction to Union Rule"; }
+
+void DisjunctionToUnionRule::apply_to(const std::shared_ptr<AbstractLQPNode>& root) const {
+  Assert(root->type == LQPNodeType::Root, "DisjunctionToUnionRule needs root to hold onto");
+
+  /**
+   * Step 1:
+   *    - Collect PredicateNodes that can be split up into multiple ones into `predicate_nodes_to_flat_disjunctions`
+   */
+  auto predicate_nodes_to_flat_disjunctions =
+      std::vector<std::pair<std::shared_ptr<PredicateNode>, std::vector<std::shared_ptr<AbstractExpression>>>>{};
+
+  visit_lqp(root, [&](const auto& sub_node) {
+    if (const auto predicate_node = std::dynamic_pointer_cast<PredicateNode>(sub_node)) {
+      const auto flat_disjunction = flatten_logical_expressions(predicate_node->predicate(), LogicalOperator::Or);
+
+      if (flat_disjunction.size() > 1) {
+        predicate_nodes_to_flat_disjunctions.emplace_back(predicate_node, flat_disjunction);
+      }
+    }
+
+    return LQPVisitation::VisitInputs;
+  });
+
+  /**
+   * Step 2:
+   *    - Split up qualifying PredicateNodes into n-1 consecutive UnionNodes and n PredicateNodes. We have to do this in
+   *      a second pass because manipulating the LQP within `visit_lqp()`, while theoretically possible, is prone to
+   *      bugs.
+   */
+  for (const auto& [predicate_node, flat_disjunction] : predicate_nodes_to_flat_disjunctions) {
+    if (flat_disjunction.size() <= 1) {
+      break;
+    }
+
+    auto previous_union_node = UnionNode::make(UnionMode::Positions);
```

Comment:
I gave this some more thought. Reliably identifying cases where the two (or more) sides are truly disjunct might take more than a handful of lines. Let's leave that for a future step. However, maybe you could benchmark how much of a difference it would make by simply replacing it and disregarding overlaps for the benchmark.

Reply:
![image](https://user-images.githubusercontent.com/12159412/63593339-15151880-c5b4-11e9-8a06-5f8624e4a81b.png)
If we use `UnionAll` instead of `UnionPositions`, TPC-H 16 can be optimized further. It does not seem to be a large potential though.

Reply:
![image](https://user-images.githubusercontent.com/12159412/63761957-48142080-c8c2-11e9-8ed7-8295cab563ff.png)
This is the same comparison if we run the `PredicateSplitUpRule` after the `DisjunctionToUnionRule`. There still is no major advantage except for TPC-H 16.